### DISCUSSION TITLE
Can I specify the context type of middlewares somehow?

### DISCUSSION THREAD

**Discussion by pgvr** [trpc/trpc#2779] [Q&A] [answered]
Hey there 👋

I have the following scenario:
two middlewares, one for authentication, one for other stuff but relies on the auth middleware to set the context.

```ts
export const authMiddleware = t.middleware(async ({ ctx, next, meta }) => {
	const session = await getSession(ctx.req)

	const identity = getIdentity(session)

	if (!identity.tenantId) {
		throw new TRPCError({ code: "UNAUTHORIZED", message: "You are not a member of a registered Tenant" })
	}

	if (meta.every) {
		requireAllPermissions(identity, meta.every)
	}
	if (meta.some) {
		requireOnePermission(identity, meta.some)
	}

	return next({
		ctx: {
			...ctx,
			identity: { ...identity } as Identity,
		},
	})
})
```

and another one where `identity` does not exist on the context.

```ts
export const anotherMiddleware = t.middleware(async ({ ctx, next, meta, rawInput }) => {
	const userId = ctx.identity.id // identity does not exist on context
}
```

Is there a way to specify the context type for middlewares? I tried looking at the generics of the middleware function but could not figure it out.

Chaining it like this also does not work because I want to use "anotherMiddleware" in several places but the chaining returns a procedure, not a middleware.

```ts
const anotherMiddleware = t.procedure.use(authMiddleware).use(({ ctx, next }) => {
         // ctx.identity is defined but this returns a procedure and not a middleware
	return next()
})
```

What I have done now is to create another "trpc"-object with a more defined Context type.

```ts
export const anotherT = initTRPC
	.context<AugmentedContext>() // AugmentedContext defines the identity to be defined
		transformer: superjson,
	})
```

If I now define the middleware based on `anotherT`, the types work but I have to make sure that the `authMiddleware` always run before the the `anotherMiddleware`. This is easy enough to ensure, but I was wondering whether this would be considered ok to do or whether there are any gotchas about having multiple "trpc"-objects.

```ts
export const anotherMiddleware = anotherT.middleware(async ({ ctx, next, meta, rawInput }) => {
	const userId = ctx.identity.id // identity does exist on context
}
```

Any help is much appreciated 🙏

**Comment by gino** [comment]
This is basically what I'm talking about in #2770 😅

**Reply by pgvr** [reply]
True, how did I not see that. I'd still be curious whether it's fine to call `initTRPC.create()` multiple times.

**Reply by gino** [reply]
No worries! I don't think so, according to the [docs](https://trpc.io/docs/v10/router#initialize-trpc). Will try to see if there are any temporary workarounds for now.

**Comment by gino** [comment, marked as answer]
The solution mentioned in #2770 should also apply to the issue that you're having.

**Reply by pgvr** [reply]
thanks for pinging me about it, much appreciated 🙏

**Reply by pencilcheck** [reply]
I'm not that smart, I don't understand how it is a solution, do you mean by chaining everything ad hoc in one file will allow typescript to infer the context type?

I kept the tutorial format where I use t.middleware() and then const new = t.procedure.use(middleware) to chain and it keeps complaining about some attribute in ctx doesn't exists

I don't really want to put everything in one file. that is not a solution to me
